rust/parser/mod.rs

```diff
@@ -20,5 +20,598 @@
+    todo!()
+}
+
+fn visit_filter(ctx: Rc<FilterContext>) -> ParserResult<Vec<UnboundVariable>> {
+    Ok((0..).map_while(|i| ctx.VAR_(i)).map(get_var).collect())
+}
+
+fn visit_sort(ctx: Rc<SortContext>) -> ParserResult<Sorting> {
+    Ok(Sorting::new(
+        (0..).map_while(|i| ctx.VAR_(i)).map(get_var).collect(),
+        &if let Some(order) = ctx.ORDER_() {
+            order.get_text()
+        } else {
+            String::from("") // FIXME
+        },
+    ))
+}
+
+fn visit_offset(_ctx: Rc<OffsetContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_limit(_ctx: Rc<LimitContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_match_aggregate(_ctx: Rc<Match_aggregateContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_aggregate_method(_ctx: Rc<Aggregate_methodContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_match_group(_ctx: Rc<Match_groupContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_definables(_ctx: Rc<DefinablesContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_definable(_ctx: Rc<DefinableContext>) -> ParserResult<Definable> {
+    todo!()
+}
+
+fn visit_patterns(ctx: Rc<PatternsContext>) -> ParserResult<Vec<Pattern>> {
+    (0..)
+        .map_while(|i| ctx.pattern(i))
+        .map(visit_pattern)
+        .collect()
+}
+
+fn visit_pattern(ctx: Rc<PatternContext>) -> ParserResult<Pattern> {
+    if let Some(var) = ctx.pattern_variable() {
+        Ok(visit_pattern_variable(var)?.into_pattern())
+    } else {
+        todo!()
+    }
+}
+
+fn visit_pattern_conjunction(_ctx: Rc<Pattern_conjunctionContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_pattern_disjunction(_ctx: Rc<Pattern_disjunctionContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_pattern_negation(_ctx: Rc<Pattern_negationContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_pattern_variable(ctx: Rc<Pattern_variableContext>) -> ParserResult<Variable> {
+    if let Some(var_thing_any) = ctx.variable_thing_any() {
+        Ok(visit_variable_thing_any(var_thing_any)?.into_variable())
+    } else if let Some(var_type) = ctx.variable_type() {
+        Ok(visit_variable_type(var_type)?.into_variable())
+    } else if let Some(var_concept) = ctx.variable_concept() {
+        Ok(visit_variable_concept(var_concept)?.into_variable())
+    } else {
+        Err(ILLEGAL_GRAMMAR.format(&[&ctx.get_text()]))
+    }
+}
+
+fn visit_variable_concept(_ctx: Rc<Variable_conceptContext>) -> ParserResult<BoundVariable> {
+    todo!()
+}
+
+fn visit_variable_type(ctx: Rc<Variable_typeContext>) -> ParserResult<TypeVariable> {
+    let mut var_type = match visit_type_any(ctx.type_any().unwrap())? {
+        Type::Variable(p) => p,
+        Type::Unscoped(p) => UnboundVariable::hidden().type_(p).into_type(),
+        _ => Err(ILLEGAL_STATE.format(&[line!().to_string().as_str()]))?,
+    };
+    for constraint in (0..).map_while(|i| ctx.type_constraint(i)) {
+        if constraint.PLAYS().is_some() {
```

Comment:
This chain of `if... else if ...` could be cleaner with a `match` guard right?

Reply:
It'd be a lot uglier
```rust
for constraint in (0..).map_while(|i| ctx.type_constraint(i)) {
    match constraint {
        constraint if constraint.PLAYS().is_some()  => {
            // ...
```
The `is_some()` here is pretty much just a boolean expression.